### DISCUSSION TITLE
Resolving unused primitive warnings for incompatible STLs

### DISCUSSION THREAD

**Discussion by 0xCoto** [thliebig/openEMS-Project#168] [Q&A Usage]
If I model 2 shapes on my CAD tool, such as these 2 cuboids "entering" into each other:

![image](https://github.com/thliebig/openEMS-Project/assets/25392776/d09d790a-968d-43ed-b243-eb070e4f6f94)

And perform a boolean subtraction (cut), I get the following shape:

<img width="735" alt="Screenshot 2023-12-20 at 12 39 32 PM" src="https://github.com/thliebig/openEMS-Project/assets/25392776/cbc700d3-ac80-4818-b2d3-0ccc13c4d204">

The problem is this: if I export each individual STL (i.e., 2 STL files), openEMS reads everything fine. However, if I perform a boolean operation like above, and try to load the resulting STL (i.e., 1 STL file), I get an "Unused primitive" warning, no matter what kind of mesh I use.

**STL file to reproduce:** [cube_cutout.stl.zip](https://github.com/thliebig/openEMS-Project/files/13726824/cube_cutout.stl.zip)

Has anyone faced issues where certain STLs not properly being read by openEMS? The same issue occurs on `libvtk7-dev` as well as `libvtk9-dev` (as far as I understand, the [STL parser](https://github.com/thliebig/CSXCAD/blob/master/src/CSPrimPolyhedronReader.cpp) is based on VTK). I'm struggling to see what the issue is with the STLs that have undergone boolean operations, as they visually to appears to look fine.

I understand that certain CAD tools may produce better boolean operations than others, and openEMS may be compatible with them, but I am trying to understand precisely where the issue lies to resolve this problem.

**Comment by aWZHY0yQH81uOYvH** [comment]
I assume the problem here is that your STL's topology is very strange and has many open edges, overlapping faces, and vertices in the middle of edges. Opening it in Blender and moving two vertices, you can see how strange it is. The normals are also inverted, but that probably doesn't matter. If it is loaded in PrusaSlicer (3D printing slicer), it reports 28 open edges and 60 repaired errors.

<img width="827" alt="cube_cutout_blender" src="https://github.com/thliebig/openEMS-Project/assets/32046476/eabe5a1f-e3e8-42a0-b6b0-dd24d553e8c1">

**Reply by aWZHY0yQH81uOYvH** [reply]
Example of a more appropriate triangulation of the same shape as generated by Blender
<img width="774" alt="cube_coutout_better" src="https://github.com/thliebig/openEMS-Project/assets/32046476/fb325cd8-1359-448a-9c67-0e0f1e4d5098">

[cube_cutout_blender.stl.zip](https://github.com/thliebig/openEMS-Project/files/13727541/cube_cutout_blender.stl.zip)

**Reply by thliebig** [reply]
openEMS can only realy handle "healthy" surface objects. That means it should be closed and have no open edges or anything nasty.

What the algorithm does is test (many many times) different coordinates (e.g. mesh edges) to be inside or outside the object.
If the coordinate is outside the bounding box it is obviously outside (simple/fast case). If not a ray from that coordinate to one surely outside is tested for intersections with the surface. If it intersects an odd number of times it is inside. If it is 0 or even it is considered outside. 
This obviously does not work well with poorly defined surfaces.

**Comment by thliebig** [comment]
Why are you not just doing the (pseudo) "boolean" operation in openEMS?  You define the base object as your material that you want and the other one as "air" material with a higher priority.

**Comment by 0xCoto** [comment]
Thank you both for your inputs - I figured out the issue and resolved it by using another tool for the boolean operations.
